File: Problems/200.number-of-islands.py
# title: Number of Islands
# timestamp: 2025-01-06 18:39:20
# problemUrl: https://leetcode.com/problems/number-of-islands/
# memory: 19.9 MB
# runtime: 241 ms

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        m, n = len(grid), len(grid[0])
        count = 0
        for k in range(m):
            for l in range(n):
                if grid[k][l] != '1':
                    continue
                q = [(k, l)]
                grid[k][l] = '2'
                count += 1
                while q:
                    i, j = q.pop()
                    for i, j in [(i-1, j), (i, j-1), (i, j+1), (i+1, j)]:
                        if 0 <= i < m and 0 <= j < n and grid[i][j] == '1':
                            q.append((i, j))
                            grid[i][j] = '2'
        return count

File: Problems/200.number-of-islands.go
// title: Number of Islands
// timestamp: 2024-04-19 22:09:27
// problemUrl: https://leetcode.com/problems/number-of-islands/
// memory: 6.1 MB
// runtime: 16 ms

func numIslands(grid [][]byte) int {
    g := make([][]int, len(grid))
    for i, line := range grid {
        g[i] = make([]int, len(line))
    }

    counter, unionNo := 1, 0
    for i, line := range grid {
        for j, cell := range line {
            if cell  == '0' {
                continue
            }
            var color int
            if i > 0 && g[i - 1][j] != 0 {
                color = g[i - 1][j]
            }
            if j > 0 && g[i][j - 1] != 0 {
                if color != 0 && color != g[i][j - 1] {
                    color = min(color, g[i][j - 1])
                    union(g, i, j)
                    unionNo++
                } else {
                    color = g[i][j - 1]
                }
            }
            if color == 0 {
                color = counter
                counter++
            }
            g[i][j] = color
        }
    }
    return counter - unionNo - 1
}

func union(g [][]int, i, j int) {
    color := min(g[i - 1][j], g[i][j - 1])
    var oldColor int
    queue := [][2]int{}
    if color == g[i - 1][j] {
        queue = append(queue, [2]int{i, j - 1})
        oldColor = g[i][j - 1]
    } else {
        queue = append(queue, [2]int{i - 1, j})
        oldColor = g[i - 1][j]
    }
    for len(queue) > 0 {
        top := queue[0]
        i, j := top[0], top[1]
        queue = queue[1:]

        g[i][j] = color

        if i > 0 && g[i - 1][j] == oldColor {
            queue = append(queue, [2]int{i - 1, j})
        }
        if j > 0 && g[i][j - 1] == oldColor {
            queue = append(queue, [2]int{i, j - 1})
        }
        if j < len(g[i]) - 1 && g[i][j + 1] == oldColor {
            queue = append(queue, [2]int{i, j + 1})
        }
        if i < len(g) - 1 && g[i + 1][j] == oldColor {
            queue = append(queue, [2]int{i + 1, j})
        }
    }
}